sdk/sdktypes/session_log_record_describe.go

```diff
@@ -0,0 +1,285 @@
+
+func (r SessionLogRecord) Describe(opts *SessionLogRecordDescribeOptions) string {
+	if opts == nil {
+		opts = DefaultSessionLogRecordDescribeOptions
+	}
+
+	m := r.read()
+
+	b := &strings.Builder{}
+
+	if opts.IncludeProcessID && m.ProcessId != "" {
+		fmt.Fprintf(b, "%s ", m.ProcessId)
+	}
+
+	if opts.IncludeTime && m.T != nil {
+		fmt.Fprintf(b, "[%s] ", r.Timestamp().Format(opts.TimeFormat))
+	}
+
+	switch {
+	case m.Print != nil:
+		fmt.Fprintf(b, "Print: %s\n", m.Print.Text)
+	case m.StopRequest != nil:
+		fmt.Fprintf(b, "Stop Request: %s\n", m.StopRequest.Reason)
+	case m.State != nil:
+		fmt.Fprint(b, "State: ")
+		describeState(opts, b, r.GetState())
+	case m.CallSpec != nil:
+		fmt.Fprint(b, "Call Spec: ")
+		describeCallSpec(opts, b, r.GetCallSpec())
+	case m.CallAttemptStart != nil:
+		fmt.Fprintf(b, "Call Attempt Start: Attempt #%d\n", m.CallAttemptStart.Num)
+	case m.CallAttemptComplete != nil:
+		fmt.Fprintln(b, "Call Attempt Complete")
+		w := opts.indent(b)
+		if m.CallAttemptComplete.IsLast {
+			fmt.Fprintln(w, "Final")
+		} else {
+			fmt.Fprintf(w, "Retry in %s\n", m.CallAttemptComplete.RetryInterval)
+		}
+		if m.CallAttemptComplete.Result.Error != nil {
+			fmt.Fprintln(w, "Error:")
+			describeError(opts, opts.indent(w), kittehs.Must1(ProgramErrorFromProto(m.CallAttemptComplete.Result.Error)))
+		} else {
+			fmt.Fprintln(w, "Return Value: ")
+			describeValue(opts, opts.indent(w), kittehs.Must1(ValueFromProto(m.CallAttemptComplete.Result.Value)))
+		}
+	default:
+		fmt.Fprintln(b, "???")
+	}
+
+	return b.String()
+}
+
+func describeCallSpec(opts *SessionLogRecordDescribeOptions, w io.Writer, s SessionCallSpec) {
+	fmt.Fprintf(w, "seq #%d, %s\n", s.Seq(), s.m.Function.Function.Name)
+
+	w = opts.indent(w)
+
+	fmt.Fprintln(w, "Function:")
+	describeFunctionValue(opts, opts.indent(w), s.Function())
+
+	for i, a := range s.m.Args {
+		fmt.Fprintf(w, "Arg #%d:\n", i)
+		describeValue(opts, opts.indent(w), kittehs.Must1(ValueFromProto(a)))
+	}
+	for k, v := range s.m.Kwargs {
+		fmt.Fprintf(w, "Arg %q: ", k)
+		describeValue(opts, opts.indent(w), kittehs.Must1(ValueFromProto(v)))
+	}
+}
+
+func describeState(opts *SessionLogRecordDescribeOptions, w io.Writer, s SessionState) {
+	fmt.Fprintf(w, "%v ", s.Type())
+
+	switch s := s.Concrete().(type) {
+	case SessionStateCreated:
+		// nop
+	case SessionStateStopped:
+		fmt.Fprintf(w, "Reason: %s", s.read().Reason)
+	case SessionStateError:
+		fmt.Fprintln(w, "")
+		describeErrorState(opts, opts.indent(w), s)
+	case SessionStateRunning:
+		if call := s.Call(); call.IsValid() {
+			fmt.Fprintln(w, call.GetFunction().Name())
+			w := opts.indent(w)
+			fmt.Fprintln(w, "Call:")
+			describeFunctionValue(opts, opts.indent(w), call)
+		} else {
+			fmt.Fprintln(w, "")
+		}
+		fmt.Fprintf(w, "%sRunID: %s\n", opts.Indent, s.read().RunId)
+	case SessionStateCompleted:
+		fmt.Fprintln(w, "")
+		describeCompletedState(opts, opts.indent(w), s)
+	default:
+		fmt.Fprintln(w, "???")
```

Comment:
Please use an informative error message that users can understand

Reply:
done